libarchive 3.7.2: fails to build with mbedtls 3.x (3.5.1)

mbedtls 2.28.6:
```
checking for mbedtls/aes.h... yes
checking for mbedtls/md.h... yes
checking for mbedtls/pkcs5.h... yes
checking for mbedtls_sha1_init in -lmbedcrypto... yes
checking support for ARCHIVE_CRYPTO_MD5_MBEDTLS... yes
checking support for ARCHIVE_CRYPTO_RMD160_MBEDTLS... yes
checking support for ARCHIVE_CRYPTO_SHA1_MBEDTLS... yes
checking support for ARCHIVE_CRYPTO_SHA256_MBEDTLS... yes
checking support for ARCHIVE_CRYPTO_SHA384_MBEDTLS... yes
checking support for ARCHIVE_CRYPTO_SHA512_MBEDTLS... yes
```
mbedtls: 3.5.1:
```
checking for mbedtls/aes.h... yes
checking for mbedtls/md.h... yes
checking for mbedtls/pkcs5.h... yes
checking for mbedtls_sha1_init in -lmbedcrypto... yes
checking support for ARCHIVE_CRYPTO_MD5_MBEDTLS... no
checking support for ARCHIVE_CRYPTO_RMD160_MBEDTLS... no
checking support for ARCHIVE_CRYPTO_SHA1_MBEDTLS... no
checking support for ARCHIVE_CRYPTO_SHA256_MBEDTLS... no
checking support for ARCHIVE_CRYPTO_SHA384_MBEDTLS... no
checking support for ARCHIVE_CRYPTO_SHA512_MBEDTLS... no
[...]
make[1]: Leaving directory '/var/tmp/paludis/build/app-arch-libarchive-3.7.2/work/libarchive-3.7.2'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_hmac_update'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_hmac_finish'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_free'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_info_from_type'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_pkcs5_pbkdf2_hmac'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_aes_setkey_enc'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_setup'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_aes_free'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_aes_crypt_ecb'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_init'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_aes_init'
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: ./.libs/libarchive.so: undefined reference to `mbedtls_md_hmac_starts'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:4765: bsdtar] Error 1
make: *** [Makefile:4027: all] Error 2
```
full build log: 
[libarchive-3.7.2_build.log](https://github.com/libarchive/libarchive/files/13491121/libarchive-3.7.2_build.log)

Migrating from Mbed TLS 2.x to Mbed TLS 3.0:
https://github.com/Mbed-TLS/mbedtls/blob/development/docs/3.0-migration-guide.md

--- a/libarchive/archive_digest.c
+++ b/libarchive/archive_digest.c
@@ -260,7 +260,11 @@ static int
 __archive_md5init(archive_md5_ctx *ctx)
 {
   mbedtls_md5_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_md5_starts(ctx) == 0)
+#else
   if (mbedtls_md5_starts_ret(ctx) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -270,7 +274,11 @@ static int
 __archive_md5update(archive_md5_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_md5_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_md5_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -279,7 +287,11 @@ __archive_md5update(archive_md5_ctx *ctx, const void *indata,
 static int
 __archive_md5final(archive_md5_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_md5_finish(ctx, md) == 0) {
+#else
   if (mbedtls_md5_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_md5_free(ctx);
     return (ARCHIVE_OK);
   } else {
@@ -431,7 +443,11 @@ static int
 __archive_ripemd160init(archive_rmd160_ctx *ctx)
 {
   mbedtls_ripemd160_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_ripemd160_starts(ctx) == 0)
+#else
   if (mbedtls_ripemd160_starts_ret(ctx) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -441,7 +457,11 @@ static int
 __archive_ripemd160update(archive_rmd160_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_ripemd160_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_ripemd160_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -450,7 +470,11 @@ __archive_ripemd160update(archive_rmd160_ctx *ctx, const void *indata,
 static int
 __archive_ripemd160final(archive_rmd160_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_ripemd160_finish(ctx, md) == 0) {
+#else
   if (mbedtls_ripemd160_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_ripemd160_free(ctx);
     return (ARCHIVE_OK);
   } else {
@@ -647,7 +671,11 @@ static int
 __archive_sha1init(archive_sha1_ctx *ctx)
 {
   mbedtls_sha1_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha1_starts(ctx) == 0)
+#else
   if (mbedtls_sha1_starts_ret(ctx) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -657,7 +685,11 @@ static int
 __archive_sha1update(archive_sha1_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha1_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_sha1_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -666,7 +698,11 @@ __archive_sha1update(archive_sha1_ctx *ctx, const void *indata,
 static int
 __archive_sha1final(archive_sha1_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha1_finish(ctx, md) == 0) {
+#else
   if (mbedtls_sha1_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_sha1_free(ctx);
     return (ARCHIVE_OK);
   } else {
@@ -915,7 +951,11 @@ static int
 __archive_sha256init(archive_sha256_ctx *ctx)
 {
   mbedtls_sha256_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha256_starts(ctx, 0) == 0)
+#else
   if (mbedtls_sha256_starts_ret(ctx, 0) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -925,7 +965,11 @@ static int
 __archive_sha256update(archive_sha256_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha256_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_sha256_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -934,7 +978,11 @@ __archive_sha256update(archive_sha256_ctx *ctx, const void *indata,
 static int
 __archive_sha256final(archive_sha256_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha256_finish(ctx, md) == 0) {
+#else
   if (mbedtls_sha256_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_sha256_free(ctx);
     return (ARCHIVE_OK);
   } else {
@@ -1155,7 +1203,11 @@ static int
 __archive_sha384init(archive_sha384_ctx *ctx)
 {
   mbedtls_sha512_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_starts(ctx, 1) == 0)
+#else
   if (mbedtls_sha512_starts_ret(ctx, 1) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -1165,7 +1217,11 @@ static int
 __archive_sha384update(archive_sha384_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_sha512_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -1174,7 +1230,11 @@ __archive_sha384update(archive_sha384_ctx *ctx, const void *indata,
 static int
 __archive_sha384final(archive_sha384_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_finish(ctx, md) == 0) {
+#else
   if (mbedtls_sha512_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_sha512_free(ctx);
     return (ARCHIVE_OK);
   } else {
@@ -1419,7 +1479,11 @@ static int
 __archive_sha512init(archive_sha512_ctx *ctx)
 {
   mbedtls_sha512_init(ctx);
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_starts(ctx, 0) == 0)
+#else
   if (mbedtls_sha512_starts_ret(ctx, 0) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -1429,7 +1493,11 @@ static int
 __archive_sha512update(archive_sha512_ctx *ctx, const void *indata,
     size_t insize)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_update(ctx, indata, insize) == 0)
+#else
   if (mbedtls_sha512_update_ret(ctx, indata, insize) == 0)
+#endif
     return (ARCHIVE_OK);
   else
     return (ARCHIVE_FATAL);
@@ -1438,7 +1506,11 @@ __archive_sha512update(archive_sha512_ctx *ctx, const void *indata,
 static int
 __archive_sha512final(archive_sha512_ctx *ctx, void *md)
 {
+#if MBEDTLS_VERSION_NUMBER > 0x03000000
+  if (mbedtls_sha512_finish(ctx, md) == 0) {
+#else
   if (mbedtls_sha512_finish_ret(ctx, md) == 0) {
+#endif
     mbedtls_sha512_free(ctx);
     return (ARCHIVE_OK);
   } else {

